cli/test_util.rs

```diff
@@ -62,16 +64,29 @@ pub fn http_server<'a>() -> HttpServerGuard<'a> {
   let g = GUARD.lock().unwrap();
 
   println!("tools/http_server.py starting...");
-  let child = Command::new("python")
+  let mut child = Command::new("python")
     .current_dir(root_path())
-    .arg("tools/http_server.py")
+    .args(&["-u", "tools/http_server.py"])
+    .stdout(Stdio::piped())
     .spawn()
     .expect("failed to execute child");
 
-  // Wait 1 second for the server to come up. TODO(ry) this is Racy.
-  std::thread::sleep(std::time::Duration::from_secs(2));
-
-  println!("tools/http_server.py ready");
+  let mut buf = [0; 1];
+  for i in 0..100 {
+    let maybe_out = child.stdout.as_mut();
+    if let Some(out) = maybe_out {
+      // http_server.py prints "ready" to stdout.
+      let r = out.read_exact(&mut buf);
+      if r.is_ok() {
+        println!("tools/http_server.py ready");
+        break
+      }
+    }
+    std::thread::sleep(std::time::Duration::from_millis(100));
+    if i == 99 {
+      panic!("tools/http_server.py failed to start");
+    }
```

Comment:
Try this code, it is race free:

```rust
  let stderr = child.stderr.as_mut().unwrap();
  let mut lines = BufReader::new(stderr).lines();
  let line = lines.next().unwrap().unwrap();
  assert!(line.starts_with("ready"));
  // Now the http server is usable.
```

Reply:
what if the server is never ready? will it throw?

Reply:
No, it will hang - but I believe there are timeouts in the rust test runner.

Reply:
I think it just says "048_media_types_jsx has been running over 60 seconds" it doesn't cancel. Though there is the one hour time out 😬

Reply:
Might need to flush the stdout in python?

Reply:
That's the -u above, in the python Command. Not sure if necessary but I added it.

Looks like your update is good! (It errors if the socket is in use.)